### ISSUE TITLE
Skopeo errors pulling images from Azure Container Registry

### ISSUE THREAD

**Issue by zhill** [anchore/anchore-engine#18]
When trying to analyze images from an Azure Container registry, the anchore-engine catalog service raises an error trying to pull the manifest and returns a 404 to the user indicating that the tag cannot be analyzed or found.

Error logs indicate the system is unable to get the manifest using skopeo:
```
[anchore_engine.auth.skopeo_wrapper/get_image_manifest_skopeo()] [WARN] CMD failed - exception: command failed: cmd=skopeo inspect --raw --tls-verify=true --creds <redacteduser>:<redactedpassword> docker://zhillregistry1.azurecr.io/testimages:centos7 exitcode=1 stdout= stderr=time="2018-01-03T03:03:55Z" level=fatal msg="unauthorized: authentication required"
```

Anchore uses [skopeo](https://github.com/projectatomic/skopeo) internally to fetch manifests and pull image layers, but skopeo fails to properly authenticate because it expects a returned bearer token to have the token content in the body as a json object with key _token_ (e.g. {"token": "...."}) as docker hub and other registries do, but Azure returns the token with key "access_key". Docker Hub returns both _token_ and _access_token_ with the same content in it's token fetch response.

Azure does support Basic Auth rather than Bearer auth, but because the Azure registry service returns a Bearer challenge in the WWW-Authenticate header of an unauthenticated request, most tools (including Skopeo) detect it and attempt to retrieve a token. But, I see no existing way to force skopeo to try basic auth and either skip the preflight auth check (a GET on the base /v2/ route) or ignore the resulting WWW-Authenticate response header, so it seems best to just fix the token handling.

As a simple check I patched skopeo (patch is a simple change at https://github.com/projectatomic/skopeo/blob/master/vendor/github.com/containers/image/docker/docker_client.go#L69)  and confirmed that with both Azure registry admin credentials as well as service principal credentials, the tool is able to pull the manifest and image content. As you can see that patch is actually in code that skopeo vendors in and should be patched upstream and pulled down.

*Given that this issue is not directly in Anchore Engine code, this issue will serve to track any upstream PRs as well as workarounds until the issue is resolved or a good workaround is found.*

**Comment by acathrow** [comment]
It also seems like Azure is at fault here by not returning the multiple authentication schemeas that it supports. Working around this in skopeo makes sense but I'll also try and log a support call into Microsoft.

**Comment by zhill** [comment]
To clarify, Azure is behaving correctly and returning a consistent API behavior. Basic Auth is an option, but token fetch works correctly and handles credentials properly. The bug is that Skopeo is looking for the wrong key in the json response to the token request so it fails to properly setup auth. The fix is to simply ensure that Skopeo can find the correct key in the token issuance response from the Azure auth service.

**Comment by acathrow** [comment]
It's odd that Azure doesn't show basic on it's list of schemas in the HTTP headers - it should if it supports basic auth.

**Comment by zhill** [comment]
Nowhere in the docs does it say that Basic works, I only discovered it by trying directly. So not sure if it is officially supported or not.

**Comment by zhill** [comment]
Looks like the base dependency for skopeo: github.com/containers/images was updated to work in commit:
https://github.com/containers/image/commit/c2b9a9b31a65406de624395141277398e249e9fa

I was literally preparing a commit to file  PR with nearly the same code.

So now just need to get skopeo to pull latest vendored dependencies and release with the new stuff.

**Comment by zhill** [comment]
There has been some activity in skopeo. The fix for auth is in master as of: https://github.com/projectatomic/skopeo/pull/468/commits/dc1567c8bc2e46054ce41891a57de774180c35c5

And that is included in the v0.1.29 tag so should be in next round of package updates for various distros. Current master branch of skopeo does work against Azure as well.

**Comment by zhill** [comment]
The latest version of skopeo available for Centos is 0.1.29 and now works with Azure registries. We'll be pulling that version into the next release of the Engine and will mark this ticket closed once we've verified it works and is released.

**Comment by zhill** [comment]
I've confirmed this is resolved in anchore-engine v0.2.2 using the anchore/anchore-engine:v0.2.2 image which has skopeo v0.1.29 installed by default.

[Docs on Azure Container Registry Auth](https://docs.microsoft.com/en-us/azure/container-registry/container-registry-authentication)

To use Azure CR w/registry admin credentials:
```
anchore-cli registry add <registryname>.azurecr.io <registryname> <admin password or password2>
anchore-cli image add <registryname>.azurecr.io/<repo>:<tag>
```

To use Azure CR w/Service Principal, [setup a principal](https://docs.microsoft.com/en-us/azure/container-registry/container-registry-auth-service-principal):
```
anchore-cli registry add <registryname>.azurecr.io <principal_id> <principal password>
anchore-cli image add <registryname>.azurecr.io/<repo>:<tag>
```

Therefore, I'm closing this issue.
